Test Failure (20171102-2210): PolicyExecutorTest.testStartTimeout has invalid timing assumption

The following error was reported by automated tests,

junit.framework.AssertionFailedError: 2017-11-03-02:51:56:134 The response did not contain "[SUCCESS]".  Full output is:"
ERROR: Caught exception attempting to call test method testStartTimeout on servlet web.PolicyExecutorServlet
java.lang.AssertionError: expected:<304398060> but was:<304773328>
	at web.PolicyExecutorServlet.testStartTimeout(PolicyExecutorServlet.java:4340)
	at componenttest.app.FATServlet.doGet(FATServlet.java:71)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.service(ServletWrapper.java:1255)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:743)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:440)
	at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1154)
	at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1004)
	at com.ibm.ws.webcontainer.servlet.CacheServletWrapper.handleRequest(CacheServletWrapper.java:76)
	at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:926)
	at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.run(DynamicVirtualHost.java:279)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink$TaskWrapper.run(HttpDispatcherLink.java:957)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.wrapHandlerAndExecute(HttpDispatcherLink.java:357)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.ready(HttpDispatcherLink.java:316)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleDiscrimination(HttpInboundLink.java:499)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleNewRequest(HttpInboundLink.java:433)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.processRequest(HttpInboundLink.java:313)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.ready(HttpInboundLink.java:284)
	at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.sendToDiscriminators(NewConnectionInitialReadCallback.java:165)
	at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.complete(NewConnectionInitialReadCallback.java:74)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.requestComplete(WorkQueueManager.java:501)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.attemptIO(WorkQueueManager.java:571)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.workerRun(WorkQueueManager.java:926)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager$Worker.run(WorkQueueManager.java:1015)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1153)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.lang.Thread.run(Thread.java:785)
". expected:<0> but was:<1>
	at componenttest.topology.utils.HttpUtils.findStringInHttpConnection(HttpUtils.java:527)
	at componenttest.topology.utils.HttpUtils.findStringInHttpConnection(HttpUtils.java:485)
	at componenttest.topology.utils.HttpUtils.findStringInUrl(HttpUtils.java:464)
	at componenttest.topology.utils.HttpUtils.findStringInReadyUrl(HttpUtils.java:436)
	at componenttest.topology.utils.HttpUtils.findStringInReadyUrl(HttpUtils.java:406)
	at componenttest.topology.utils.FATServletClient.runTest(FATServletClient.java:81)
	at componenttest.custom.junit.runner.SyntheticServletTest.invokeExplosively(SyntheticServletTest.java:40)
	at componenttest.custom.junit.runner.FATRunner$1.evaluate(FATRunner.java:175)
	at componenttest.custom.junit.runner.FATRunner$2.evaluate(FATRunner.java:316)
	at componenttest.custom.junit.runner.FATRunner.run(FATRunner.java:152)

This test case doesn't allow for the possibility that when two threads attempt to abort a task at the same time, and one of the threads can have successfully set the state and be finishing up the abort processing, the other thread can return (unable to abort) before the abort processing has been completed, in which case the reported elapsed queue time can change after that point, as the first thread records the timestamp.  This getElapsedQueueTime interface allows for this sort of trivial fluctuation near the point at which the task is ending, in stating that "No guarantee is made that the same exact value will be returned for subsequent invocations".  The test case cannot expect to reliably make this assert and must be updated.

--- a/dev/com.ibm.ws.threading_policy_fat/test-applications/basicfat/src/web/PolicyExecutorServlet.java
+++ b/dev/com.ibm.ws.threading_policy_fat/test-applications/basicfat/src/web/PolicyExecutorServlet.java
@@ -4368,7 +4368,10 @@ public class PolicyExecutorServlet extends FATServlet {
         assertTrue(future.isDone());
         assertFalse(future.isCancelled());
         assertEquals(0, future.getElapsedRunTime(TimeUnit.NANOSECONDS));
-        assertEquals(queueTimeNS, future.getElapsedQueueTime(TimeUnit.NANOSECONDS));
+        boolean sameQueueTime = false;
+        for (long start = System.nanoTime(); !sameQueueTime && System.nanoTime() - start < TIMEOUT_NS; Thread.sleep(200))
+            sameQueueTime = queueTimeNS == future.getElapsedQueueTime(TimeUnit.NANOSECONDS);
+        assertTrue(sameQueueTime);
 
         // Additional testing for the measured run time of the blocker task
         assertTrue(blockerFuture.get(TIMEOUT_NS, TimeUnit.NANOSECONDS));

